### ISSUE TITLE
Reopened: library maintenance (#296)

### ISSUE THREAD

**Issue by leamas** [Rasbats/shipdriver_pi#338]
This is basically about reopening #296. The basic problem stated there is that it is more or less impossible to maintain multiple, totally independent copies of libraries used in multiple projects. 

And that's just how it is. I'm convinced that some libs already have diverged, and it will only get worse over time.

I also still think #298 basically was sound. However, it came in absolutely wrong time just before the release of 3.0. For that reason it was right to reject it. Still, in  a longer perspective this is probably the direction in which we should move. But in smaller steps, taking developer concerns about changed procedures into account.

**Comment by Rasbats** [comment]
>giving some time to adjust to new workflows.

Totally agree. Just about to lead someone through the building process. The learning curve is already steep enough.

**Comment by rgleason** [comment]
For example know that plugingl has several versions in Sean's plugins.
Also I know of two versions of nmea0183,  I tried to get them together but that effort failed.

**Comment by Rasbats** [comment]
> Have I understood your position correctly?

Yes ... I want a plugin repo to have all the libraries it needs, as it stands, without having to fetch them from another repo. If there was a mechanism to add the libraries in the same way as _update-templates_ that would work for me. I can see the benefits of a store of libraries which we use as a standard.

More important for me is having a piDC that can be employed across a number of plugins. 

My EU AIS work is moving into a proof of concept stage, with live test transmissions. Time is going to be short.

**Comment by rgleason** [comment]
I thought you were going to use opencpn-lib for standard libs from opencpn?
Leaving libs for a plugins atypical?
Then the use could be known and controled. 
Update-templates would update opencpn-libs only after most of its libs moved into opencpn-lib?

**Comment by leamas** [comment]
@jongough:

> If there was a mechanism to add the libraries in the same way as update-templates that would work for me.

My thinking is based on something like this:
  - We update _update-templates_ so after running it the new opencpn-lilbs directory is populated and ready to use without further actions. 
  - _update_templates_ also  automatically updates opencpn-libs to latest version.
  - After cloning a plugin the opencpn-libs will be in a sort of empty state. This can be handled in three ways:
    - By using the `clone --recursive`  option which populates opencpn-libs without further actions
    - Or, using  ` update-templates -l`  which likewise populates the opencpn-libs  directory so it's  ready to use.
    - ` update-templates -l`  is just a shorthand for `git -C opencpn-libs pull origin master`   which also can be used.

In summary,  opencpn-libs is fully usable and populated after at most one single command, and in the normal usecase using _update_templates_ there is no need to do anything at all. 

Seems this OK?

**Comment by Rasbats** [comment]
@leamas : This would pull into opencpn-libs all the libraries. A lot of extra libs in the dev repo if only, say, one is needed. 

This might be getting a little complicated when all that is needed is to update the libs already in place to the latest version. nmea is used in a number of ways and is a prime candidate for standardizing but things will break and need fixing. tinyxml should now be updated to tinyxml2. Again a lot of coding to get things updated.  Is it all worthwhile, if the plugin just works?

**Comment by leamas** [comment]
> This would pull into opencpn-libs all the libraries. A lot of extra libs in the dev repo if only, say, one is needed.

Yes  (so what?)

> This might be getting a little complicated when all that is needed is to update the libs already in place to the latest version.

No. Currently, update-templates treats _libs/_ the same way as Plugin.cmake and build-deps -- it's the plugin's private business. Nothing here is updated in any way, and as reported versions of things which probably should be kept uniform has diverged. 

Another aspect is that shipdriver by definition is a sort of "minimal" plugin using a very small set of libraries.

> nmea is used in a number of ways and is a prime candidate for standardizing but things will break and need fixing

This is in fact a good argument for having a central library. For when we fix things there, it will be automatically deployed to the other plugins over time. Fixing it in opencpn-libs is basically the same thing  as fixing  it anyway, you basically just change files, build test and commit (the exact commit procedure is a little different).

>  Is it all worthwhile, if the plugin just works?

Perhaps not from the perspective from a specific plugin.

However, when you look at the complete opencpn universe of main program, testplugin plugins and shipdriver plugins I think it's obvious that we will gain a lot by solving the problems once and for all in a central plugin repository compared to fixing things in each place separately.

Please note the transition path here:  for each new plugin we introduce a new directory opencpn-libs. plugins still has a choice of using libs there or in the existing libs/. More compllicated plugins with many libraries will use both alternatives, a mix of "private" libs in _libs/_ and some centrally maintained in _opencpn-libs_

**Comment by rgleason** [comment]
As the libs get manually replaced with standard opencpn-libs, the authors can add the standard library that is needed, run the necessary tests and make necessary changes to code. 

Then perhaps update-templates will update those particular opencpn-lib directories only?  So as not to confuse or bloat plugin code?

**Comment by leamas** [comment]
> As the libs get manually replaced with standard opencpn-libs, 

Will not happen. What happens is that a new directory opencpn-libs/ is added. Each plugin then makes decisions about using plugins from libs/ or opencpn-libs/ (assuming it exists in opencpn-libs, otherwise libs/ is the only option)

**Comment by rgleason** [comment]
I think that is what I wrote.
"As the libs get manually replaced with standard opencpn-libs, the authors can add the standard library that is needed, run the necessary tests and make necessary changes to code."

**Comment by rgleason** [comment]
BTW, I didn't understand the term "bootstrapping". I had no clue. Is this something common in programming?

**Comment by leamas** [comment]
> BTW, I didn't understand the term "bootstrapping". I had no clue. Is this something common in programming?

Yes.

**Comment by rgleason** [comment]
Does the plugin author manually keep the opencpn-libs up to date with Opencpn?

**Comment by leamas** [comment]
No, normally fixed when applying `update-templates`, can also be done `using update-templates -L`

**Comment by leamas** [comment]
After d0472c9 some follow-ups this is now done. Closing
